lib/httpd.c

```diff
@@ -0,0 +1,176 @@
+/**
+ * @file
+ * @brief Simple HTTP server
+ * @date 16.04.12
+ * @author Ilia Vaprol
+ * @author Anton Kozlov
+ * 	- CGI related changes
+ * @author Andrey Golikov
+ * 	- Linux adaptation
+ */
+
+#include <assert.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/wait.h>
+#include <stdio.h>
+#include <stdint.h>
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#include "emhttp_lib.h"
+
+
+#define BUFF_SZ   (1024)
+
+struct client_info {
+	struct sockaddr ci_addr;
+	socklen_t ci_addrlen;
+	int ci_sock;
+	int ci_index;
+
+	const char *ci_basedir;
+};
+
+static char httpd_g_inbuf[BUFF_SZ];
+
+
+static int httpd_read_http_header(int sk, char *buf, size_t buf_sz) {
+	const char pattern[] = "\r\n\r\n";
+	char *pb;
+	int offset;
+	int header_len;
+
+	if (buf_sz <  sizeof(pattern)) {
+		return -EINVAL;
+	}
+
+	if (0 > read(sk, buf, 4)) {
+		return -errno;
+	}
+
+	pb = buf + 4;
+
+	for (header_len = 4; header_len < buf_sz; header_len++) {
+		if (0 == memcmp(pb - 4, pattern, 4)) {
+			break;
+		}
+		if (0 > read(sk, pb, 1)) {
+			return -errno;
+		}
+		pb++;
+	}
+
+	*pb = '\0';
+
+	return header_len;
+}
+
+int httpd_build_request(struct client_info *cinfo, struct http_req *hreq, char *buf, size_t buf_sz) {
+	int nbyte;
+
+	nbyte = httpd_read_http_header(cinfo->ci_sock, buf, buf_sz - 1);
+	if (nbyte < 0) {
+		printf("can't read from client socket: %s", strerror(errno));
+		return -errno;
+	}
+
+	memset(hreq, 0, sizeof(*hreq));
+	if (NULL == http_parse_request(buf, hreq)) {
+		printf("can't parse request");
+		return -EINVAL;
+	}
+
+	return nbyte;
+}
+
+static int httpd_header(const struct client_info *cinfo, const char *msg, int msg_size) {
+	int cbyte;
+	char *buf;
+	int buf_sz = sizeof(buf);
+
+	buf = http_get_text_response_header();
+
+	cbyte = strlen(buf);
+
+	if (0 > write(cinfo->ci_sock, buf, cbyte)) {
+		return -errno;
+	}
+
+	if (0 > write(cinfo->ci_sock, msg, msg_size)) {
+		return -errno;
+	}
+
+	return 0;
+}
+
+static void httpd_client_process(struct client_info *cinfo) {
+	struct http_req hreq;
+	int err;
+
+	if (0 > (err = httpd_build_request(cinfo, &hreq, httpd_g_inbuf, sizeof(httpd_g_inbuf)))) {
+		printf("can't build request: %s", strerror(-err));
+	}
+
+	printf("method=%s uri_target=%s uri_query=%s",
+			   hreq.method, hreq.uri.target, hreq.uri.query);
+
+
+	httpd_header(cinfo, "!!!!!!", sizeof("!!!!!!"));
+}
+
+void *httpd_thread(void *arg) {
+	int host;
+
+	struct sockaddr_in inaddr;
+	const int family = AF_INET;
+
+	inaddr.sin_family = AF_INET;
+	inaddr.sin_port= htons(80);
```

Comment:
Listening port should be rather passed from `arg` based on device configuration than constant 80.
For example, for `single_thing`, it should be `8888`.
```
webthing_server_run(&thing, 1, "lamp", 8888);
```